closed/src/java.base/share/classes/java/security/CRIUConfigurator.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * ===========================================================================
+ * (c) Copyright IBM Corp. 2022, 2022 All Rights Reserved
+ * ===========================================================================
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * IBM designates this particular file as subject to the "Classpath" exception
+ * as provided by IBM in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * ===========================================================================
+ */
+
+package java.security;
+
+import java.lang.reflect.Method;
+import java.util.Hashtable;
+import java.util.Properties;
+
+import sun.security.action.GetPropertyAction;
+import sun.security.jca.ProviderList;
+import sun.security.jca.Providers;
+
+/**
+ * Configures the security providers when in CRIU mode.
+ */
+public final class CRIUConfigurator {
+    private static Properties systemProps;
+    /** Stores the old security providers (position, name) */
+    private static Hashtable<String, String> oldProviders = new Hashtable<String, String>();
+    /** Tracing for CRIUSEC */
+    private static final boolean debug = Boolean.parseBoolean(GetPropertyAction.privilegedGetProperty
+            ("enable.j9internal.checkpoint.security.api.debug", "false"));
+
+    /**
+     * Removes the usual security providers and adds the CRIU security provider.
+     *
+     * @param props the system properties
+     */
+    public static void setCRIUSecMode(Properties props) {
+        systemProps = props;
+
+        for(Object keyObj : props.keySet()) {
+            String key = (String) keyObj;
+            if(key.startsWith("security.provider")) {
+                oldProviders.put(key, props.getProperty(key));
+            }
+        }
+        for(String provider : oldProviders.keySet()) {
+            props.remove(provider);
+        }
+        props.put("security.provider.1", "com.ibm.security.criu.CRIUSECProvider");
+
+        if (debug) {
+            System.out.println("CRIUSEC added and all other security providers removed.");
+        }
+    }
+
+    /**
+     * Removes the CRIU security provider and adds the usual security providers back.
+     */
+    public static void setCRIURestoreMode() {
+        Security.removeProvider("CRIUSEC");
+        systemProps.remove("security.provider.1");
```

Comment:
How can we be sure that `security.provider.1` corresponds with `CRIUSEC`?

Reply:
The method `setCRIURestoreMode` is called after `setCRIUSecMode`, in which we put `CRIUSEC` as `security.provider.1`. Should I put a check anyway?

Reply:
I think a comment would be sufficient to help future readers of this code.

Reply:
alright, thank you! added a comment